pkg/strategy/grid2/recover.go

```diff
@@ -0,0 +1,305 @@
+
+	if s.GridProfitStats.InitialOrderID != 0 {
+		// existiting strategy, need recover
+	} else if len(openOrdersOnGrid) != 0 {
+		// open orders on grid is not 0. this strategy is existing, need recover
+	} else {
+		// initial order id may be new strategy or lost data in redis, so we need to check trades + open orders
+		// if there are open orders or trades, we need to recover
+		trades, err := historyService.QueryTrades(ctx, s.Symbol, &types.TradeQueryOptions{
+			// from 1, because some API will ignore 0 last trade id
+			LastTradeID: 1,
+			// if there is any trades, we need to recover.
+			Limit: 1,
+		})
+
+		if err != nil {
+			return errors.Wrapf(err, "check if strategy exists error when query trades")
+		}
+
+		if len(trades) == 0 {
+			s.logger.Info("new strategy, no need to recover")
+			return nil
+		}
+	}
+
+	s.logger.Infof("start to recover")
+	if err := s.recoverWithOpenOrdersByScanningTrades(ctx, historyService, openOrdersOnGrid); err != nil {
+		return errors.Wrap(err, "grid recover error")
+	}
+
+	return nil
+}
+
+func (s *Strategy) recoverWithOpenOrdersByScanningTrades(ctx context.Context, historyService types.ExchangeTradeHistoryService, openOrdersOnGrid []types.Order) error {
+	if s.orderQueryService == nil {
+		return fmt.Errorf("orderQueryService is nil, it can't get orders by trade")
+	}
+
+	// set grid
+	grid := s.newGrid()
+	s.setGrid(grid)
+
+	// add open orders to active order book
+	s.addOrdersToActiveOrderBook(openOrdersOnGrid)
+
+	expectedOrderNums := s.GridNum - 1
+	openOrdersOnGridNums := int64(len(openOrdersOnGrid))
+	s.debugLog("open orders nums: %d, expected nums: %d", openOrdersOnGridNums, expectedOrderNums)
+	if expectedOrderNums == openOrdersOnGridNums {
+		// no need to recover
+		return nil
+	} else if expectedOrderNums < openOrdersOnGridNums {
+		return fmt.Errorf("amount of grid's open orders should not > amount of expected grid's orders")
+	}
+
+	// 1. build pin-order map
+	pinOrdersOpen, err := s.buildPinOrderMap(grid.Pins, openOrdersOnGrid)
+	if err != nil {
+		return errors.Wrapf(err, "failed to build pin order map with open orders")
+	}
+
+	// 2. build the filled pin-order map by querying trades
+	pinOrdersFilled, err := s.buildFilledPinOrderMapFromTrades(ctx, historyService, pinOrdersOpen)
+	if err != nil {
+		return errors.Wrapf(err, "failed to build filled pin order map")
+	}
+
+	// 3. get the filled orders from pin-order map
+	filledOrders := pinOrdersFilled.AscendingOrders()
+	numFilledOrders := len(filledOrders)
+	if numFilledOrders == int(expectedOrderNums-openOrdersOnGridNums) {
+		// nums of filled order is the same as Size - 1 - num(open orders)
+	} else if numFilledOrders == int(expectedOrderNums-openOrdersOnGridNums+1) {
+		filledOrders = filledOrders[1:]
+	} else {
+		return fmt.Errorf("not reasonable num of filled orders")
+	}
+
+	// 4. verify the grid
+	if err := s.verifyFilledGrid(s.grid.Pins, pinOrdersOpen, filledOrders); err != nil {
```

Comment:
do we have log in the verifyFilledGrid method?

Reply:
yes